### ISSUE TITLE
CRIU GC Implementation 

### ISSUE THREAD

**Issue by RSalman** [eclipse/omr#6888]
This issue outlines the work being done around GC to support CRIU. It is intended to document discussions, progress and priorities for the GA release. The current effort for this release is focused on supporting the standard collectors (i.e., performance and correctness with gencon, optthruput and optavgpause gc policies). The Balanced and Metronome GC policies require further investigation/implementation and are not exhaustively detailed in this issue.

**Areas focused on for GA:**
1. GC Dispatcher Thread Pool Tuning (thread count adjustments)
2. Reinitialization  of GC components for restore (specifically to compensate for thread count changes)
3. Verbose GC
4. Post Restore Options Parsing

**Other areas to consider:**
- Heap resizing
- GC stats at different phases and implications on heuristics
- NUMA configuration _(Non-standard policy)_
- Metronome Scheduler _(Non-standard policy)_

## 1. GC (Dispatcher) Thread Pool Tuning
- [X] **Adjust GC Thread Count between startup, checkpoint and restore (contract/expand thread pool for C/R)**
Optimize GC footprint observed when utilizing CRIU. Tests have shown excessive number of GC threads to have a significant impact on checkpoint image footprint and restore times. Contracting the GC thread pool (thread shutdown) during the checkpoint phase and expanding it (thread startup) during restore phase has shown to be a better strategy.

The following GC threading optimization has been implemented: 
- Startup: CRIU checkpoint run thread count init remains unchanged , i.e. GC thread count set same manner traditional VM startup.
- Checkpoint: thread count is reduced to 4 for the checkpoint image (or user specified by `-XX:CheckpointGCThreadCount=`)
- Restore:  new thread count is determined for the restore machine (computed in the same manner as initial VM startup, based on H/W). Restore thread count is bounded by startup thread count.

- [X] **Unbounded Thread Pool Expansion**
The dispatcher thread pool is allocated at VM startup, it is sized based on the startup thread count, i.e the  checkpoint machine hardware. This results in a max capacity/bound and ultimately limits the restore thread count. The restore machine may support a higher thread count than the max imposed by checkpoint machine. Test must be done to determine the implications of reallocating the thread pool past initial VM allocation and load test for correctness (i.e GC components dependency on initial thread count needs to be understood).

- [ ] **Single threaded GC mode for startup and checkpoint run/phase**
Eventually, we should explore the idea of starting the checkpoint run with a single GC thread and make thread adjustment only at restore time. With this approach, checkpoint image size and restore times are expected to be reduced as GC footprint is minimized, i.e GC structures that scale based on thread count are minimized. Work in area `#2` must be completed before this can be explored. That is, thread count compensations are necessary to avoid performance penalties if GC components are inited based on single threaded mode.

- [ ] **Async thread startup (lazy init)**
Thread startup for restore is synchronous, it blocks the restore thread from proceeding until all GC threads are started and are in ready state in the dispatcher. Restore times can be reduced If the thread startup routines can be made asynchronous. This is an experimental idea which requires further discussion and prototyping to explore viability. Low priority and currently not planned. 
 
## 2. Compensate for Thread Count Change
There are side effects to adjusting the GC thread count. Various GC components/structures are initialized based on the initial thread count at VM startup. Traditional use of the VM/GC assumes that the total thread count doesn't exceed the # threads inited on startup, this is not the case with CRIU. Hence, a change in thread count must be compensated for by adjusting components that were inited based on a different thread count. Such components must consume the new thread count at restore time. Testing has proven this is for performance reasons and not required for correctness.

GC components  that require adjustment are enumerated below. These are discovered through code analysis, primarily by following code paths that consume GC thread counts. The following keywords (var names, function names etc) related to GC thread count were considered:
- `gcThreadCount`
- `threadCountMaximum`
- `threadCount`
- `activeThreadCount`
- `WorkerID`
- GC Thread Iteration

Not all the code paths are significant, majority can be disregarded because they consume the thread count at gc time. We are mostly interested in components that consume the thread count at initialization/startup time:

|Status | Component|Comments 
|--- | --- | ---|
|<ul><li>- [ ] </li></ul>| Sweep | Reinit sweepChunkSize and resize chunk pool based on new sweepChunkSize.
|<ul><li>- [ ] </li></ul>| Configuration, PacketList and CopyScanCacheList | Reallocate splitFreeListSplitAmount, packetListSplit and cacheListSplit
|<ul><li>- [ ] </li></ul>| WorkPackets | initialPacketCount
|<ul><li>- [ ] </li></ul>| HeapRegionDescriptorExtensions (Obj Lists) | objectListFragmentCount, _unfinalizedObjectLists, _ownableSynchronizerObjectLists, _continuationObjectLists & _referenceObjectLists
|<ul><li>- [ ] </li></ul>| Monitor Table | initializeMonitorTable / monitorTableListPool
|<ul><li>- [ ] </li></ul>| String Table |  tableCount == GC Thread Count 

### Verbose GC
See https://github.com/eclipse-openj9/openj9/issues/16601

> Log filenames can be reinitialized at restore time, this is an approach that we discussed for GC. Depending on the use case, this approach may (or may not) result in separate log files for the two checkpoint/restore runs.
> 
> Different scenarios with this approach:
> 
> * By default, VerboseGC filename is set to: `verbosegc.%Y%m%d.%H%M%S.%pid.txt`. Once this is re-evaluated at restore, we'll have a different filename and hence 2 separate log files.
> * If the user explicitly sets the filename for the checkpoint run and not restore run, the existing checkpoint log file will be reused/overwritten for the restore run. This is the same behaviour as running a VM multiple times (with the same verbosegc filename set).
> * When explicitly setting filenames, two separate files can kept if the user specifics a different filename for the restore run
> 
> I think this aligns to the existing behaviours. This behaviour can extend to rotating log files as well. When the rotating log name is shared between checkpoint and restore, I think we should simply start logging at the first file in the rotation.

Immediate issues with verbose GC + CRIU were addressed by: https://github.com/eclipse/omr/pull/6850

- [x] Close/Re-open verbose file streams at C/R
- [ ] Re-evaluate file name
- [ ] Checkpoint/Restore identifier stanza
- [ ] Rotating log files: file pointer on restore 
- [ ] Init stanza needs to be reprinted for non-file stream outputs
- [ ] Parse Verbose GC related opts

### Post Restore Options Parsing
GC defaults need to be reinitialized at restore time and cmd line options need to be considered. Parameters that are machine specific should be reset based on the restore machine.

**Comment by RSalman** [comment]
@amicic @dmitripivkine @tajila @vijaysun-omr

**Comment by RSalman** [comment]
Verbose GC logging must be properly initialized and reconfigured, taking into account both restore options and initial verbose configuration at startup. There are multiple configuration scenarios possible depending on whether logging is enabled/disabled (pre/post restore) and the options provided at each phase:

- Restore options will override the startup verbose logging configuration.  
- In the absence on verbose logging restore options, startup options will determine the logging configuration. 
- Default file names will be reinitialized. 
- Currently not possible to disable verbose logging, if enabled at startup

The following table exhausts the various scenarios and the expected behaviour, used as test cases: 

![image](https://user-images.githubusercontent.com/9648788/223540224-d28d6ed1-7c9c-4efc-a67f-c3e015377218.png)

**Comment by RSalman** [comment]
There are also rotating log files which have been left out of the table. They are essentially an extension of named logs and follow the same rule expect the file rotation log is also reset to the initial log in the rotation.

**Comment by amicic** [comment]
After hearing opinion from other teams (core VM and JIT) that also may deal with same problem, for the case where we already have explicitly stated VGC log file name on checkpoint side and we did not specify anything on restore side (or explicitly specified the same name), we should not overwrite the existing file, but actually append to it.

**Comment by dmitripivkine** [comment]
I think we should keep in mind possible case when files created pre-restore are not available at post-restore any more. Do we need extra logic to handle this case?

**Comment by RSalman** [comment]
Proposing the following explicit Verbose GC restore run identifier: `<attribute name="CRIU Restore" value="true" />`. This can show up in the `<initialized>` stanza

```
<initialized id="1" timestamp="2023-03-08T11:49:38.738">
  <attribute name="gcPolicy" value="-Xgcpolicy:gencon" />
  <attribute name="maxHeapSize" value="0xfa210000" />
  <attribute name="initialHeapSize" value="0x800000" />
  <attribute name="compressedRefs" value="true" />
  <attribute name="compressedRefsDisplacement" value="0x0" />
  <attribute name="compressedRefsShift" value="0x3" />
  <attribute name="pageSize" value="0x1000" />
  <attribute name="pageType" value="not used" />
  <attribute name="requestedPageSize" value="0x1000" />
  <attribute name="requestedPageType" value="not used" />
  <attribute name="gcthreads" value="5" />
  <attribute name="gcthreads Concurrent Mark" value="1" />
  <attribute name="packetListSplit" value="3" />
  <attribute name="cacheListSplit" value="3" />
  <attribute name="splitFreeListSplitAmount" value="3" />
  <attribute name="numaNodes" value="0" />
...
  <attribute name="CRIU Restore" value="true" />   ** CRIU Identifier 
...
  <system>
    <attribute name="physicalMemory" value="16785920000" />
    <attribute name="numCPUs" value="8" />
...
```

Outside of a restore run, the identifier can be omitted rather than present a "false" value

**Comment by RSalman** [comment]
>I think we should keep in mind possible case when files created pre-restore are not available at post-restore any more. Do we need extra logic to handle this case?

The existing fileOpen logic handles this case, a new file is created when the file is not available.

**Comment by RSalman** [comment]
I agree, how about `Restore Snapshot`.. or `Restored Snapshot`?

**Comment by RSalman** [comment]
In terms of GC threading options, we should finalize the behaviour when `Xgcthreads` is specified at VM startup. 

Right now, GC will shutdown threads irrespective of the startup options. I think this is fine, its required to minimize footprint / restore times. However, the question is what should be done when starting threads back up at restore time...  current implementation, we ignore the startup -Xgcthreads  option and either look to 
- `-Xgcthreads` provided at restore time, or 
- The restore machine supported thread count

The current implementation can be changed to respect initial startup -Xgcthreads option when a thread count is not explicitly set in restore args

**Comment by amicic** [comment]
for restore VGC identifier, any new one that you suggested is ok with me

**Comment by RSalman** [comment]
Currently, pool expansion is limited to restore and pool contraction is limited to checkpoint. We should eventually make expansion/contraction independent of the phase. This should be done for completeness to lift restrictions from `Xgcthreads` option. There is limited value in this, hence it is fairly low priority.
https://github.com/eclipse-openj9/openj9/pull/16883#discussion_r1137988773

**Comment by RSalman** [comment]
Internal GC lists need to be adjusted to optimize for the new thread count at restore. Depending on whether the thread count increases or decreases, the lists need to be further split into sublists, or be merged into fewer sublists. Merging sublists is non-trivial and can be expensive, it requires compressing/connecting sublists which can be complex and messy when multiple lists are involved (may involve walking the entire sublist). Doing this at restore can be be detrimental to restore startup times. **A decision was made to compromise and ignore list merge (with decrease in thread count) and only handle list split (increase in thread count).** 

We don't expect this to be an issue because the average case is not impacted i.e., no detrimental effects from excessive list splitting for majority of GC cycles. Additionally, this is relevant only when checkpoint is taken with a very large thread count and restore is done with much smaller thread count.

**Background**
Internal GC lists are split into multiple sublists to reduce contention with multiple GC threads. An optimal number of sublists is required, dividing sublist excessively can be an issue as lists can be partially filled (or left empty) and consequently, threads may incur extra overhead (acquire list locks, walk lists etc). This is mostly true during the end of GC cycle and not so much an issue at other times (i.e., doesn’t impact the average case). Currently, a list is split into a sublist for every 8 threads, This is done for 
1. Memory Pool (https://github.com/eclipse/omr/pull/6946)
3. Scavenger Cache lists (https://github.com/eclipse/omr/pull/6970)
4. Work Packet lists (https://github.com/eclipse/omr/pull/6966)

**Comment by RSalman** [comment]
Verbose GC reports split amount, this exception will only need to be clear for restore verbose reporting.

**Comment by RSalman** [comment]
>  Currently, a list is split into a sublist for every 8 threads. This is done for  
> 1. Memory Pool (https://github.com/eclipse/omr/pull/6946)

In terms of Memory Pool, it's not simple to split pool lists during restore  f the lists weren't already split at the time of the VM's initial startup (need at least 2 sublists to split further at restore).Specifically, when initial VM startup thread count doesn't require split lists ( < 8 startup threads), but restore thread count ( > 8 restore threads) does. This is because the pool type setup/allocated depends on startup split count. A split pool is allocated when startup split count > 1, default AddressOrderedList Pool otherwise, which doesn't support split lists. Ideally, split pool should behave as non-split pool with 1 list, however this is not the case with how the memory pools are implemented. I'm currently looking into converting non split memory pool into a split pool for this specific case.

**Comment by amicic** [comment]
I think the scenario when restore side tries to decrease the number of GC threads ends up with default GC threads on restore machine. It should not completely ignore it, but to the best closest to it.

For example, if checkpoint VM had 8 threads, and restore tries 4, and the restore machine has 16 h/w threads, we should run with 8 (what is closer to requested 4), and not 16 (what is default for the machine).
